### PR TITLE
Consequences of module monomorphisms

### PR DESCRIPTION
**Pull request by Taneb**
Progess:
- [x] Left semimodules
- [x] Left modules
- [x] Right semimodules
- [x] Right modules
- [x] Bisemimodules
- [x] Bimodules
- [x] Semimodules
- [x] Modules

Creating this PR now to get feedback as soon as possible. I'm intending to add to it as I have the energy.

### CODE REVIEW

File: src/Algebra/Module/Morphism/LeftSemimoduleMonomorphism.agda [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+open import Algebra.Module.Bundles.Raw
+open import Algebra.Module.Morphism.Structures
+
+module Algebra.Module.Morphism.LeftSemimoduleMonomorphism
+  {r a b ℓ₁ ℓ₂} {R : Set r} {M₁ : RawLeftSemimodule R a ℓ₁} {M₂ : RawLeftSemimodule R b ℓ₂} {⟦_⟧}
+  (isLeftSemimoduleMonomorphism : IsLeftSemimoduleMonomorphism M₁ M₂ ⟦_⟧)
+  where
+
+open IsLeftSemimoduleMonomorphism isLeftSemimoduleMonomorphism
+open RawLeftSemimodule M₁ renaming (Carrierᴹ to A; _≈ᴹ_ to _≈ᴹ₁_; _+ᴹ_ to _+ᴹ_; 0ᴹ to 0ᴹ₁; _*ₗ_ to _*ₗ_)
```
Comment (JacquesCarette) [inline src/Algebra/Module/Morphism/LeftSemimoduleMonomorphism.agda:18]: Have you considered name these `M` and `N` instead, and then using qualified names on local modules instead of all these renames?

I've experimented with both styles myself, and rather grew to prefer prefixes over rename+subscripts. But it's definitely a style issue!

Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/LeftSemimoduleMonomorphism.agda:18]: Two things about this:
* I agree with @JacquesCarette and @MatthewDaggitt in their advocacy for private modules and qualified names, over renaming; the additions in [`FreeMagma`](https://github.com/agda/agda-stdlib/pull/1962) benefited enormously from this nudge;
* the existing boilerplate in `Algebra.Morphism.Structures` unfortunately adopts the 'older' style, so any attempt to migrate ought to be accompanied by a global (and exhaustingly boring) refactoring...

So, I'd be tempted to let this stand as it is for now, provided we post a `style-guide` issue on the need to do this eventually?

**Comment by jamesmckinna** [comment]
Can we have a global comment about the purpose of the PR? 

IIUC, a `...ModuleMonomorphism` allows you to induce a particular `Is...Module` structure on the source `Raw...Module` bundle by pulling back along the monomorphism... ie lifting `On._on_` from `Relation.Binary`/`Function` to a particular algebraic structure? It feels as though we ought (sic) to be able to abstract over this in some more generic way? Or then again, from experience with stdlib, maybe not... yet. But I can't help thinking that all the `Algebra.*.Monomorphism` modules might better be refactored... somehow? Apologies if this remark is also out-of-scope for the PR at hand, and should be a `style-guide`/`library-design` or `README.Design.Hierarchy` issue instead?

**Comment by Taneb** [comment]
@jamesmckinna I'm not sure what you're asking for there. I don't think `Relation.Binary.Construct.On` significantly helps us here and I can't think of a way of doing this more generically without a colossal amount of risky work. But this is an obvious missing piece, and useful for my next goal of defining `Ideal`s.

**Comment by jamesmckinna** [comment]
OK @Taneb I'll try to rephrase my question, but happy to do so later downstream.
But in any case, a link back to your #2076 would have been helpful (to me, at least) for context...

**Comment by jamesmckinna** [comment]
I think I'd been trying to understand what you were doing in (something like) the following terms (a lifting property from `AbGrp` to `R-Module` arising from a monadic situation):
* if you have an `AbelianGroupMonomorphism` between the underlying `AbelianGroup`s of two modules, *and* it's a module *homo*morphism, then you actually have a module *mono*morphism; 
* hence (cf. `Algebra.Structures.Biased` etc.), it suffices to supply 'simpler' data in order to actually have 'richer' structure, moreover automatically... ie to construct an ideal of a ring R it suffices to supply 'an Abelian subgroup which is closed under multiplication'...

But it seems as though what you have is doing something else here?

**Comment by jamesmckinna** [comment]
As for doing things generically, yes, your point is well-made. But I think the 'lifting' idea above is something to look out for generally when thinking about 'concrete categories' such as `Group` over `Setoid` and `R-Mod` over `AbGrp`...? (perhaps the `agda-categories` gang can give chapter and verse here?)

**Comment by JacquesCarette** [comment]
I think @jamesmckinna is simply asking: can you explain in plain English what you are trying to achieve here? It's not clear just from the code itself.

**Comment by Taneb** [comment]
What I'm trying to achieve here: I want a quick and easy way to prove that I have a module given that I have a raw module and a monomorphism to some other module.

**Comment by jamesmckinna** [comment]
Re: the discussion above about `renaming` directives vs. `private` module defs, one reason for the gymnastics in the first place might be our failure to adopt the related idioms discussed in the recent https://github.com/agda/agda/issues/7118 and related material (eg. [Conal's message](https://agda.zulipchat.com/#narrow/stream/238741-general/topic/Instance.20resolution.20failure.20in.20Agda.202.2E6.2E4) on the Zulip...)? Still not sure if I quite understand what is at stake there, but anything to try to reduce the amount of boilerplate involved with the `Algebra.*` hierarchy (as I am discovering on #2296 ...)

**Comment by jamesmckinna** [comment]
Looking at the parametrisation of the various constructions, I can't help but see this as further argument in favour of #2252 , but in any case without having to address that issue head-on, you might still consider making all of these definitions inside anonymous modules which would allow you to avoid the [DRY] anti-pattern of `open`ing a locally-defined module using `let` in the *type* of an definition, and then having to repeat yourself with a corresponding `where` clause in the body ...?

But taking #2252 more seriously, your operations coul all be parametrised on `RawX` bundles, rather than their flattened versions, and then `open` them locally in the types/definitions as above in order to obtain the exact symbols you require for the `IsX` assumptions...?

If I had more time today, I'd re-draft a specimen instance for comparison... but that will have to wait for now.

**Comment by Taneb** [comment]
@jamesmckinna I've made things a bit simpler with modules. Thanks for the suggestion! I don't know how much #2252 will help, though

**Review by jamesmckinna — APPROVED** [review]
This a PROVISIONAL approve review.
Everything looks good, but there seems to be an excess of `open`ing of structures which end up not being used... so will nit-pick in what follows...

File: src/Algebra/Module/Morphism/BimoduleMonomorphism.agda [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  private
+    R-ring : Ring _ _
+    R-ring = record { isRing = R-isRing }
+
+    S-ring : Ring _ _
+    S-ring = record { isRing = S-isRing }
+
+    module R = IsRing R-isRing
+    module S = IsRing S-isRing
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/BimoduleMonomorphism.agda:61]: Eg as `module R = Ring (record  isRing = isRing })` and then use `R.<fieldname>` thereafter?

File: src/Algebra/Module/Morphism/RightModuleMonomorphism.agda [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+------------------------------------------------------------------------
+-- Structures
+
+module _ {ℓr} {_≈_ : Rel R ℓr} {_+_ _*_ -_ 0# 1#} (R-isRing : IsRing _≈_ _+_ _*_ -_ 0# 1#) where
+
+  private
+    R-ring : Ring _ _
+    R-ring = record { isRing = R-isRing}
+  open IsRing R-isRing
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/RightModuleMonomorphism.agda:50]: Either way, whitespace missing in `record { isRing = R-isRing }`!

**Comment by jamesmckinna** [comment]
Sorry that this has taken so long!

File: src/Algebra/Module/Morphism/LeftModuleMonomorphism.agda [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+------------------------------------------------------------------------
+-- Structures
+
+module _ {ℓr} {_≈_ : Rel R ℓr} {_+_ _*_ -_ 0# 1#} (R-isRing : IsRing _≈_ _+_ _*_ -_ 0# 1#) where
+
+  private
+    R-ring : Ring _ _
+    R-ring = record { isRing = R-isRing }
+  open IsRing R-isRing
```
[the remark this answers is not in the document]
Comment (Taneb) [reply src/Algebra/Module/Morphism/LeftModuleMonomorphism.agda:51]: As a matter of personal taste, I preferred to `open` records where I didn't think that there was any ambiguity

[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/LeftModuleMonomorphism.agda:51]: OK, then.

File: src/Algebra/Module/Morphism/LeftModuleMonomorphism.agda [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  isLeftModule
+    : IsLeftModule R-ring N._≈ᴹ_ N._+ᴹ_ N.0ᴹ N.-ᴹ_ N._*ₗ_
+    → IsLeftModule R-ring M._≈ᴹ_ M._+ᴹ_ M.0ᴹ M.-ᴹ_ M._*ₗ_
+  isLeftModule isLeftModule = record
+    { isLeftSemimodule = isLeftSemimodule isSemiring NN.isLeftSemimodule
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/LeftModuleMonomorphism.agda:57]: Ditto. (there may be no ambiguity, but there may be additional cognitive load on the reader?)

File: src/Algebra/Module/Morphism/ModuleMonomorphism.agda [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+------------------------------------------------------------------------
+-- Structures
+
+module _ {ℓr} {_≈_ : Rel R ℓr} {_+_ _*_ -_ 0# 1#} (R-isCommutativeRing : IsCommutativeRing _≈_ _+_ _*_ -_ 0# 1#) where
+
+  private
+    R-commutativeRing : CommutativeRing _ _
+    R-commutativeRing = record { isCommutativeRing = R-isCommutativeRing }
+
+  open IsCommutativeRing R-isCommutativeRing
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/ModuleMonomorphism.agda:44]: As above.

File: src/Algebra/Module/Morphism/ModuleMonomorphism.agda [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+  isModule
+    : IsModule R-commutativeRing N._≈ᴹ_ N._+ᴹ_ N.0ᴹ N.-ᴹ_ N._*ₗ_ N._*ᵣ_
+    → IsModule R-commutativeRing M._≈ᴹ_ M._+ᴹ_ M.0ᴹ M.-ᴹ_ M._*ₗ_ M._*ᵣ_
+  isModule isModule = record
+    { isBimodule = isBimodule isRing isRing NN.isBimodule
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/ModuleMonomorphism.agda:50]: Ditto.

File: src/Algebra/Module/Morphism/RightModuleMonomorphism.agda [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  isRightModule
+    : IsRightModule R-ring N._≈ᴹ_ N._+ᴹ_ N.0ᴹ N.-ᴹ_ N._*ᵣ_
+    → IsRightModule R-ring M._≈ᴹ_ M._+ᴹ_ M.0ᴹ M.-ᴹ_ M._*ᵣ_
+  isRightModule isRightModule = record
+    { isRightSemimodule = isRightSemimodule isSemiring NN.isRightSemimodule
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/RightModuleMonomorphism.agda:56]: Ditto.

File: src/Algebra/Module/Morphism/SemimoduleMonomorphism.agda [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+------------------------------------------------------------------------
+-- Structures
+
+module _ {ℓr} {_≈_ : Rel R ℓr} {_+_ _*_ 0# 1#} (R-isCommutativeSemiring : IsCommutativeSemiring _≈_ _+_ _*_ 0# 1#) where
+
+  private
+    R-commutativeSemiring : CommutativeSemiring _ _
+    R-commutativeSemiring = record { isCommutativeSemiring = R-isCommutativeSemiring }
+
+  open IsCommutativeSemiring R-isCommutativeSemiring
```
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/SemimoduleMonomorphism.agda:65]: As above.

File: src/Algebra/Module/Morphism/BimoduleMonomorphism.agda [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  isBimodule
+    : IsBimodule R-ring S-ring N._≈ᴹ_ N._+ᴹ_ N.0ᴹ N.-ᴹ_ N._*ₗ_ N._*ᵣ_
+    → IsBimodule R-ring S-ring M._≈ᴹ_ M._+ᴹ_ M.0ᴹ M.-ᴹ_ M._*ₗ_ M._*ᵣ_
```
Comment (jamesmckinna) [inline src/Algebra/Module/Morphism/BimoduleMonomorphism.agda:65]: I see... here. OK!

File: src/Algebra/Module/Morphism/BimoduleMonomorphism.agda (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamesmckinna) [reply src/Algebra/Module/Morphism/BimoduleMonomorphism.agda:61]: Yes, I see now. I was too hasty, sorry.

File: src/Algebra/Module/Morphism/LeftSemimoduleMonomorphism.agda [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+module Algebra.Module.Morphism.LeftSemimoduleMonomorphism
+  {r a b ℓ₁ ℓ₂} {R : Set r} {M₁ : RawLeftSemimodule R a ℓ₁} {M₂ : RawLeftSemimodule R b ℓ₂} {⟦_⟧}
+  (isLeftSemimoduleMonomorphism : IsLeftSemimoduleMonomorphism M₁ M₂ ⟦_⟧)
+  where
+
+open IsLeftSemimoduleMonomorphism isLeftSemimoduleMonomorphism
+private
+  module M = RawLeftSemimodule M₁
+  module N = RawLeftSemimodule M₂
```
Comment (jamesmckinna) [inline src/Algebra/Module/Morphism/LeftSemimoduleMonomorphism.agda:20]: Elsewhere you use `M` and `N` where here you have `M₁` and `M₂`?
Does it make sense to (try to) standardise on one or the other throughout this PR?

**Review by jamesmckinna — APPROVED** [review]
